server/src/main/java/org/candlepin/model/ProductCurator.java

```diff
@@ -334,15 +273,346 @@ else if (boolAttrs != null && boolAttrs.contains(attr.getName())) {
+                pa.setProduct(entity);
+                this.validateAttributeValue(pa);
+            }
+        }
+
+        if (entity.getProductContent() != null) {
+            for (ProductContent pc : entity.getProductContent()) {
+                pc.setProduct(entity);
+            }
+        }
+
+        // TODO: Add more reference checks here.
+
+        return entity;
+    }
+
+    /**
+     * Updates the product references currently pointing to the original product to instead point to
+     * the updated product for the specified owners.
+     *
+     * @param current
+     *  The current product other objects are referencing
+     *
+     * @param updated
+     *  The product other objects should reference
+     *
+     * @param owners
+     *  A collection of owners for which to apply the reference changes
+     *
+     * @return
+     *  a reference to the updated product
+     */
+    protected Product updateOwnerProductReferences(Product current, Product updated,
+        Collection<Owner> owners) {
+        // Impl note:
+        // We're doing this in straight SQL because direct use of the ORM would require querying all
+        // of these objects and HQL refuses to do any joining (implicit or otherwise), which
+        // prevents it from updating collections backed by a join table.
+        // As an added bonus, it's quicker, but we'll have to be mindful of the memory vs backend
+        // state divergence.
+
+        // TODO:
+        // These may end up needing to change if we hit the MySQL 65k-element limitation on IN.
+
+        Session session = this.currentSession();
+        Set<String> ownerIds = new HashSet<String>();
+
+        for (Owner owner : owners) {
+            ownerIds.add(owner.getId());
+        }
+
+        // Owner products
+        String sql = "UPDATE cp2_owner_products SET product_uuid = ?1 " +
+            "WHERE product_uuid = ?2 AND owner_id IN (?3)";
+
+        int opCount = session.createSQLQuery(sql)
+            .setParameter("1", updated.getUuid())
+            .setParameter("2", current.getUuid())
+            .setParameterList("3", ownerIds)
+            .executeUpdate();
+
+        log.debug("{} owner-product relations updated", opCount);
+
+        // Activation key products
+        sql = "UPDATE cp2_activation_key_products SET product_uuid = ?1 " +
+            "WHERE product_uuid = ?2 AND key_id IN (SELECT id FROM cp_activation_key WHERE owner_id IN (?3))";
+
+        int akCount = session.createSQLQuery(sql)
+            .setParameter("1", updated.getUuid())
+            .setParameter("2", current.getUuid())
+            .setParameterList("3", ownerIds)
+            .executeUpdate();
+
+        log.debug("{} activation keys updated", akCount);
+
+        // Installed products
+        sql = "UPDATE cp2_installed_products SET product_uuid = ?1 " +
+            "WHERE product_uuid = ?2 AND consumer_id IN (SELECT id FROM cp_consumer WHERE owner_id IN (?3))";
```

Comment:
I realize you have a note above regarding possibly reaching the IN clause limitation, but, I'm particularly concerned about the IN clause here since the sub select is targeting consumer ids in multiple owners. The result could easy blow that limit if I'm reading the query right.